src/coreclr/src/jit/lclvars.cpp

```diff
@@ -4080,8 +4125,20 @@ void Compiler::lvaComputeRefCounts(bool isRecompute, bool setSlotNumbers)
                     case GT_STORE_LCL_VAR:
                     case GT_STORE_LCL_FLD:
                     {
-                        const unsigned lclNum = node->AsLclVarCommon()->GetLclNum();
-                        lvaTable[lclNum].incRefCnts(weight, this);
+                        LclVarDsc* varDsc = lvaGetDesc(node->AsLclVarCommon());
+                        // If this is an EH var, use a zero weight for defs, so that we don't
+                        // count those in our heuristic for register allocation, since they always
+                        // must be stored, so there's no value in enregistering them at defs; only
+                        // if there are enough uses to justify it.
+                        if (varDsc->lvLiveInOutOfHndlr && !varDsc->lvDoNotEnregister &&
+                            ((node->gtFlags & GTF_VAR_DEF) != 0))
+                        {
+                            varDsc->incRefCnts(0, this);
```

Comment:
Is there some other way to accomplish this? I never like to see us bias this sort of accounting if we can avoid it.

Say down the road we want to use weighted ref counts for other optimizations -- for instance trying to allocate the most frequently accessed locals at small FP or SP offsets -- we'd want these weighted ref counts to reflect our best model of reality.

Reply:
This is tricky because of the way we overload the ref counts for use by optimization and register allocation - the register allocator uses the weight to determine the value of allocating a register.
Note that this is only the `IsLir()` path.